core/services/pipeline/graph.go

```diff
@@ -182,19 +95,90 @@ func (n *TaskDAGNode) Attributes() []encoding.Attribute {
 	return r
 }
 
-func (n *TaskDAGNode) outputs() []*TaskDAGNode {
-	var nodes []*TaskDAGNode
-	ns := n.g.From(n.ID())
-	for ns.Next() {
-		nodes = append(nodes, ns.Node().(*TaskDAGNode))
+type Pipeline struct {
+	Tasks  []Task
+	tree   *Graph
+	Source string
+}
+
+func (p *Pipeline) UnmarshalText(bs []byte) (err error) {
+	parsed, err := Parse(string(bs))
+	if err != nil {
+		return err
+	}
+	*p = *parsed
+	return nil
+}
+func (p *Pipeline) MinTimeout() (time.Duration, bool, error) {
+	var minTimeout time.Duration = 1<<63 - 1
+	var aTimeoutSet bool
+	for _, t := range p.Tasks {
+		if timeout, set := t.TaskTimeout(); set && timeout < minTimeout {
+			minTimeout = timeout
+			aTimeoutSet = true
+		}
 	}
-	return nodes
+	return minTimeout, aTimeoutSet, nil
 }
 
-func unwrapGraphNodes(nodes graph.Nodes) []*TaskDAGNode {
-	var out []*TaskDAGNode
-	for nodes.Next() {
-		out = append(out, nodes.Node().(*TaskDAGNode))
+func Parse(text string) (*Pipeline, error) {
+	g := NewGraph()
+	err := g.UnmarshalText([]byte(text))
+
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Pipeline{
+		tree:   g,
+		Tasks:  make([]Task, 0, g.Nodes().Len()),
+		Source: text,
 	}
-	return out
+
+	// toposort all the nodes: dependencies ordered before outputs. This also does cycle checking for us.
+	nodes, err := topo.SortStabilized(g, nil)
+
+	// we need a temporary mapping of graph.IDs to positional ids after toposort
+	ids := make(map[int64]int)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to topologically sort the graph, cycle detected")
+	}
+
+	// use the new ordering as the id so that we can easily reproduce the original toposort
+	for id, node := range nodes {
+		node, is := node.(*GraphNode)
+		if !is {
+			panic("unreachable")
+		}
+
+		if node.dotID == InputTaskKey {
+			return nil, errors.Errorf("'%v' is a reserved keyword that cannot be used as a task's name", InputTaskKey)
+		}
+
+		task, err := UnmarshalTaskFromMap(TaskType(node.attrs["type"]), node.attrs, id, node.dotID)
+		if err != nil {
+			return nil, err
+		}
+
+		// re-link the edges
+		for inputs := g.To(node.ID()); inputs.Next(); {
+			from := p.Tasks[ids[inputs.Node().ID()]]
+
+			from.Base().outputs = append(from.Base().outputs, task)
+			task.Base().inputs = append(task.Base().inputs, from)
+		}
+
+		// This is subtle: g.To doesn't return nodes in deterministic order, which would occasionally swap the order
+		// of inputs, therefore we manually sort. We don't need to sort outputs the same way because these appends happen
+		// in p.Task order, which is deterministic via topo.SortStable.
+		sort.Slice(task.Base().inputs, func(i, j int) bool {
```

Comment:
Ordering is now deterministic but I hope the explanation here is clear on why we only sort inputs.